source/extensions/filters/http/common/jwks_fetcher.cc

```diff
@@ -106,19 +146,75 @@ class JwksFetcherImpl : public JwksFetcher,
   Upstream::ClusterManager& cm_;
   bool complete_{};
   JwksFetcher::JwksReceiver* receiver_{};
-  const envoy::config::core::v3::HttpUri* uri_{};
+
   Http::AsyncClient::Request* request_{};
 
+  Tracing::Span* parent_span_{};
+
+  const envoy::extensions::filters::http::jwt_authn::v3::RemoteJwks& remote_jwks_;
+  const envoy::config::core::v3::HttpUri& uri_;
+
+  Envoy::Event::Dispatcher& dispatcher_;
+
+  Envoy::BackOffStrategyPtr backoff_strategy_;
+
+  uint32_t num_retries_;
+
+  uint32_t retries_remaining_;
+
+  Envoy::Random::RandomGeneratorImpl random_;
+
   void reset() {
     request_ = nullptr;
     receiver_ = nullptr;
-    uri_ = nullptr;
+    parent_span_ = nullptr;
+
+    // truncated backoff back to 0 retries attempted.
+    retries_remaining_ = num_retries_;
+
+    if (backoff_strategy_) {
+      // backoff strategy : back to initial (small) delay
+      backoff_strategy_->reset();
+    }
+  }
+
+  void retryFetch(JwksFetcher::JwksReceiver::Failure reason) {
+
+    // cant' fetch() if receiver isn't null... ( for example after a reset() )
+    auto* receiver = receiver_;
+    receiver_ = nullptr;
+
+    if (backoff_strategy_) {
+      if (retries_remaining_-- > 0) {
+
+        auto retry_ms = std::chrono::milliseconds(backoff_strategy_->nextBackOffMs());
+
+        ENVOY_LOG(warn, "retrying after {} milliseconds backoff", retry_ms.count());
+
+        auto backoff_timer =
```

Comment:
I think you need to store this timer into class member,  otherwise when out of scope, it is freed.

Reply:
makes sense. most likely I'll get all of that right early next week.

Reply:
This is straightforwards to implement, but when I do, 
the `//test/extensions/filters/http/jwt_authn:filter_integration_test` test fails.
It is using a libevent based timer instead of a mock ( which I guess is the desired way to perform an integration test), but now the `dispatcher_->createTimer()` call inside `JwksFetcherImpl`'s constructor complains fails its "am I thread-safe" assertion inside `source/common/event/dispatcher_impl.cc`'s 

```
TimerPtr DispatcherImpl::createTimer(TimerCb cb) {
  ASSERT(isThreadSafe());
  return createTimerInternal(cb);
}
```

maybe you can shed some light ?

Reply:
dispatcher is per-thread, we have to use the dispatch for that worker thread. I think you may have pass in the wrong dispatcher for authenticator.  authenticator is created per-thread.

Reply:
One way to do that is  to add Dispatcher& into [ContextImpl](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/jwt_authn/verifier.cc#L34)

and pass decoder_callbacks_->dispatcher()  defined [here](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/jwt_authn/filter.h#L40)

Reply:
done.